make_iterator_from_pointer.hpp

```diff
@@ -0,0 +1,256 @@
+        using pointer = const T*;
+        using reference = const T&;
+        using iterator_concept = std::contiguous_iterator_tag;
+
+    public:
+        wrap_pointer_to_const_iterator() = default;
+        wrap_pointer_to_const_iterator(const wrap_pointer_to_const_iterator &) = default;
+
+        wrap_pointer_to_const_iterator(wrap_pointer_to_const_iterator &&) = default;
+
+        wrap_pointer_to_const_iterator &operator=(const wrap_pointer_to_const_iterator &) = default;
+
+        wrap_pointer_to_const_iterator &operator=(wrap_pointer_to_const_iterator &&) = default;
+
+        ~wrap_pointer_to_const_iterator() = default;
+        reference operator*() const;
+
+        pointer operator->() const;
+
+        wrap_pointer_to_const_iterator &operator++();
+
+        wrap_pointer_to_const_iterator operator++(int);
+
+        wrap_pointer_to_const_iterator &operator--();
+
+        wrap_pointer_to_const_iterator operator--(int);
+
+        wrap_pointer_to_const_iterator &operator+=(difference_type);
+
+        wrap_pointer_to_const_iterator operator+(difference_type) const;
+
+        wrap_pointer_to_const_iterator &operator-=(difference_type);
+
+        wrap_pointer_to_const_iterator operator-(difference_type) const;
+
+        wrap_pointer_to_const_iterator operator[](difference_type) const;
+
+        bool operator==(const wrap_pointer_to_const_iterator &) const;
+
+    protected:
+        pointer _ptr = nullptr;
+
+        explicit wrap_pointer_to_const_iterator(pointer);
+
+        template<typename t,std::size_t N>
+        friend class ::lghazarosyan::array;
+
+    };
+
+    template<class T>
+    class wrap_pointer_to_iterator;
+
+    template<class T>
+    class wrap_pointer_to_iterator<const T*>;
+
+    template<class T>
+    class wrap_pointer_to_iterator<T*> : public wrap_pointer_to_const_iterator<T*> {
+
+    public:
+        using value_type = T;
+        using difference_type = std::ptrdiff_t;
+        using iterator_category = std::contiguous_iterator_tag;
+        using pointer = T*;
+        using reference = T&;
+        using iterator_concept = std::contiguous_iterator_tag;
+    public:
+        reference operator*() const;
+
+        pointer operator->() const;
+
+        wrap_pointer_to_iterator &operator++();
+
+        wrap_pointer_to_iterator operator++(int);
+
+        wrap_pointer_to_iterator &operator--();
+
+        wrap_pointer_to_iterator operator--(int);
+
+        wrap_pointer_to_iterator operator[](difference_type) const;
+
+        wrap_pointer_to_iterator &operator+=(difference_type);
+
+        wrap_pointer_to_iterator operator+(difference_type) const;
+
+        wrap_pointer_to_iterator &operator-=(difference_type);
+
+        wrap_pointer_to_iterator operator-(difference_type) const;
+
+    private:
+        wrap_pointer_to_iterator(const wrap_pointer_to_const_iterator<T*> &);
+
+        explicit wrap_pointer_to_iterator(pointer);
+
+        template<typename t,std::size_t N>
```

Comment:
are you gonna add each class using this as friend?

Reply:
Yea, because this is not for users, I only use this to make writing my classes easier (the iterator part)

Reply:
Is there an easier way? to complete this part? I mean not to write friend for each type we are adding?